File: langgraphjs/agents/agentic/state.ts
import { Annotation } from "@langchain/langgraph";
import { BaseMessage } from "@langchain/core/messages";

// Define state schema for the workflow
export const AgentState = Annotation.Root({
  // Planner related state
  input: Annotation<string>({
    reducer: (x, y) => y ?? x,
    default: () => "",
  }),
  plan: Annotation<{ description: string; substeps: string[] }[]>({
    reducer: (x, y) => x.concat(y),
    default: () => [],
  }),
  pastSteps: Annotation<Record<string, any>[]>({
    reducer: (x, y) => x.concat(y),
    default: () => [],
  }),
  response: Annotation<string>({
    reducer: (x, y) => y ?? x,
    default: () => "",
  }),
  sources: Annotation<string[]>({
    reducer: (x, y) => y ?? x ?? [],
    default: () => [],
  }),
  messages: Annotation<BaseMessage[]>({
    reducer: (x, y) => x.concat(y),
    default: () => [],
  }),
  next: Annotation<string>({
    reducer: (x, y) => y ?? x ?? "END",
    default: () => "END",
  }),
  instruction: Annotation<string>({
    reducer: (x, y) => y ?? x ?? "",
    default: () => "",
  })
});

File: src/state.py
from langchain_core.messages import BaseMessage
from typing_extensions import TypedDict, Annotated
from typing import List
from langgraph.graph import END

class Step(TypedDict):
  description: str
  substeps: list[str]

class State(TypedDict):
    input: Annotated[str, {
      "reducer": lambda x, y: y if y is not None else x, 
      "default": lambda: ""
    }]
    plan: Annotated[List[Step], {
      "reducer": lambda x, y: x + y, 
      "default": lambda: []
    }]
    pastSteps: Annotated[list[dict[str, any]], {
      "reducer": lambda x, y: x + y, 
      "default": lambda: []
    }]
    response: Annotated[str, {
      "reducer": lambda x, y: y if y is not None else x, 
      "default": lambda: ""
    }]
    sources: Annotated[list[str], {
      "reducer": lambda x, y: y if y is not None else x if x is not None else [], 
      "default": lambda: []
    }]
    messages: Annotated[list[BaseMessage], {
      "reducer": lambda x, y: x + y, 
      "default": lambda: []
    }]
    next: Annotated[str, {
      "reducer": lambda x, y: y if y is not None else x if x is not None else END, 
      "default": lambda: END
    }]
    instruction: Annotated[str, {
      "reducer": lambda x, y: y if y is not None else x if x is not None else "", 
      "default": lambda: ""
    }]
